### ISSUE TITLE
Inconsistent newlines for execution output

### ISSUE THREAD

**Issue by Mierdin** [StackStorm/st2#3855]
I thought this was documented in another issue but I can't find it, so recording here. Sorry if duplicate.

For the last few months I've noticed that some executions respect newlines, and others do not (and in turn, record the literal `\n` in its place).

For long outputs like in our CI where the st2 bootstrap script is used, this makes troubleshooting **awful**.

It doesn't even seem to be specific to certain runners. For instance, both actions below use `python-script`, yet the `stderr` for the first execution renders newline characters, and the second does not.

```
vagrant@st2vagrant:~$ st2 run packs.setup_virtualenv packs=napalm
......................
id: 5a07a4f0c4da5f054cf2fe22
status: succeeded
parameters:
  packs:
  - napalm
result:
  exit_code: 0
  result: 'Successfuly set up virtualenv for the following packs: napalm'
  stderr: 'st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Setting up virtualenv for pack "napalm" (/opt/stackstorm/packs/napalm)

    st2.actions.python.SetupVirtualEnvironmentAction: INFO     Virtualenv path "/opt/stackstorm/virtualenvs/napalm" doesn''t exist

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Creating virtualenv for pack "napalm" in "/opt/stackstorm/virtualenvs/napalm"

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Creating virtualenv in "/opt/stackstorm/virtualenvs/napalm" using Python binary "/opt/stackstorm/st2/bin/python"

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Running command "/opt/stackstorm/st2/bin/virtualenv -p /opt/stackstorm/st2/bin/python --always-copy /opt/stackstorm/virtualenvs/napalm" to create virtualenv.

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Installing base requirements

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Installing requirement six>=1.9.0,<2.0 with command /opt/stackstorm/virtualenvs/napalm/bin/pip install six>=1.9.0,<2.0.

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Installing pack specific requirements from "/opt/stackstorm/packs/napalm/requirements.txt"

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Installing requirements from file /opt/stackstorm/packs/napalm/requirements.txt with command /opt/stackstorm/virtualenvs/napalm/bin/pip install -U -r /opt/stackstorm/packs/napalm/requirements.txt.

    st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Virtualenv for pack "napalm" successfully created in "/opt/stackstorm/virtualenvs/napalm"

    '
  stdout: ''
vagrant@st2vagrant:~$ st2 run napalm.ping hostname=192.168.16.25 destination=192.168.16.25
.
id: 5a07a546c4da5f054cf2fe25
status: failed
parameters:
  destination: 192.168.16.25
  hostname: 192.168.16.25
result:
  exit_code: 1
  result: None
  stderr: "Traceback (most recent call last):\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2common/runners/python_action_wrapper.py\", line 259, in <module>\n    obj.run()\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2common/runners/python_action_wrapper.py\", line 155, in run\n    output = action.run(**self._parameters)\n  File \"/opt/stackstorm/packs/napalm/actions/ping.py\", line 10, in run\n    with self.get_driver(**std_kwargs) as device:\n  File \"/opt/stackstorm/packs/napalm/actions/lib/action.py\", line 37, in get_driver\n    found_device = self.find_device_from_config(hostname, driver, credentials, port)\n  File \"/opt/stackstorm/packs/napalm/actions/lib/action.py\", line 88, in find_device_from_config\n    raise Exception(message)\nException: Configuration Error: Please provide a pack config and ensure it's loaded  with `st2ctl reload --register-configs`\n"
  stdout: ''
```

I don't recall this being as bad of a problem as it is now, and I'd like to spend some time understanding how/when these newline characters are interpreted and give it some TLC to ensure this behavior is much more predictable than it is now.

Any effort against this issue should consider all action runners, and all potential output fields.

**Comment by nmaludy** [comment]
+1

This actually bit us this week. The escaped newline char resulted in us doing some find/replace statements to fix these inputs.

**Comment by armab** [comment]
I remember @LindsayHill noticed the same recently too.

Feels like a bug/regression in some recent versions.

**Comment by LindsayHill** [comment]
This is a related one https://github.com/StackStorm/st2/issues/3602

There's definitely inconsistencies between the different runners and the different output fields (CLI, st2web, etc)

**Comment by Mierdin** [comment]
I believe I have identified the problem. I noticed this was only happening with execution results at the CLI so I looked at the commit history for the execution formatter, and found [this change](https://github.com/StackStorm/st2/commit/f7f2824f97adba39c321d91b761b0ca976fa5331#diff-06e6161212caee40a9c4c6bb4b96249c) around the time I remembered seeing this behavior. /cc @lakshmi-kannan 

The culprit for this particular behavior is the call to `yaml.safe_dump`. It took me a while to realize exactly what was going on, but it's shown in [this gist](https://gist.github.com/Mierdin/baf741e6e412f6bdf29238e3e73be769). Basically, `safe_dump` seems to have trouble with newlines followed by spaces. This is probably some arcane rule somewhere I'm not aware of (`json.dumps` seems to do the same thing) but basically, if the newline is followed by a space (i.e. `\n `) then it gets converted to `\\n` by the dump. Simply removing the space results in a `\n\n`. Since the escape is escaped in the former, this screws up formatting by not actually rendering the newline characters, and simply printing them in the string as `\n`.

Before https://github.com/StackStorm/st2/pull/3511, this was addressed by running the final output through `strutils.unescape` before returning the value. This removed the first escape slash, and when printing the result, the newline character would be honored and the output would be not terrible. However, since #3511 moved this call before the `yaml.safe_dump` call, those extra escapes persisted in the string.

The good news is that it should be an easy fix. Simply adding a second call to `strutils.unescape` before returning `output`, as it was done before, fixes the problem right away (I also keep the current call before `yaml.safe_dump` to preserve the spirit of #3511. I will open a PR after I've properly tested a few things. I also think I can fix https://github.com/StackStorm/st2/issues/3602 while I'm here, so I'll open a PR once I've finished with this exercise.
